PDF export fails in production VS Code extension: webview.printToPDF unavailable

## Summary

PDF export fails in the production VS Code extension package `nimblesite.typediagram` version `0.5.0`. The command composes HTML successfully, then fails because `webview.printToPDF` is unavailable in the current VS Code runtime.

## Impact

Production users cannot export PDFs from the TypeDiagram extension in this runtime.

## Observed behavior

VS Code shows an error notification:

```text
TypeDiagram: PDF export failed -- Error: [PDF-PRINT] webview.printToPDF is not available in this VS Code runtime. Requires the Electron-based desktop VS Code.
```

The extension logs show the export was invoked and HTML composition completed before the print step failed.

## Log excerpt

```jsonl
2026-04-23T03:35:40.811Z [INFO] extension activating {"scope":"activate","version":"0.5.0","extensionPath":"<user-home>/.vscode/extensions/nimblesite.typediagram-0.5.0"}
2026-04-23T03:35:40.811Z [INFO] activate returning extendMarkdownIt contribution {"scope":"activate"}
2026-04-23T03:35:40.811Z [INFO] called by VS Code markdown preview {"scope":"extendMarkdownIt","callCount":1,"alreadyWarm":false}
2026-04-23T03:35:40.827Z [INFO] plugin installed on markdown-it instance {"scope":"md-plugin","hadPreviousFence":true}
2026-04-23T03:35:40.864Z [INFO] warmup complete, triggering markdown.preview.refresh {"scope":"extendMarkdownIt","elapsedMs":52}
2026-04-23T03:35:40.864Z [INFO] markdown.preview.refresh resolved {"scope":"extendMarkdownIt"}
2026-04-23T03:35:47.157Z [INFO] export-pdf invoked {"scope":"export-pdf","uri":"file://<workspace>/docs/plans/github-issues-lsp-mcp-vsix-plan.md"}
2026-04-23T03:35:47.168Z [INFO] composed HTML {"scope":"export-pdf","fenceCount":0,"htmlLength":28686,"diagnostics":0,"elapsedMs":11}
2026-04-23T03:35:47.330Z [ERROR] export-pdf failed {"scope":"export-pdf","err":"Error: [PDF-PRINT] webview.printToPDF is not available in this VS Code runtime. Requires the Electron-based desktop VS Code."}
```

## Expected behavior

PDF export should work in the production extension, or the command should detect unsupported VS Code runtimes before attempting export and present a supported fallback or clear preflight error.

## Environment

- Extension: `nimblesite.typediagram`
- Version: `0.5.0`
- Failure timestamp: `2026-04-23T03:35:47.330Z`
- Source: production VS Code extension install

## Notes

The failure message says this requires Electron-based desktop VS Code, but this happened from the installed production extension. The export path needs runtime capability detection and/or a fallback export implementation for unsupported runtimes.

--- a/coverage-thresholds.json
+++ b/coverage-thresholds.json
@@ -16,10 +16,10 @@
       "lines": 96.71
     },
     "packages/web": {
-      "statements": 95.43,
+      "statements": 97.64864864864865,
       "branches": 95.86,
-      "functions": 97.63,
-      "lines": 95.43
+      "functions": 97.93617021276596,
+      "lines": 97.62306368330465
     },
     "packages/vscode": {
       "statements": 99,

--- a/packages/vscode/src/export-pdf.ts
+++ b/packages/vscode/src/export-pdf.ts
@@ -21,6 +21,7 @@ export interface ExportPdfDeps {
   readonly readFile: (uri: vscode.Uri) => Promise<Uint8Array>;
   readonly writeFile: (uri: vscode.Uri, data: Uint8Array) => Promise<void>;
   readonly uriWithPath: (base: vscode.Uri, newPath: string) => vscode.Uri;
+  readonly isPrintToPdfRuntimeSupported?: () => boolean;
   readonly createWebviewPanel: (
     viewType: string,
     title: string,
@@ -182,6 +183,8 @@ export function composeHtml(mdSource: string, opts: { theme: Theme; title: strin
 // ---------------------------------------------------------------------------
 
 const PRINT_LOAD_TIMEOUT_MS = 15_000;
+const UNSUPPORTED_RUNTIME_MESSAGE =
+  "[PDF-PRINT] webview.printToPDF is not available in this VS Code runtime. Requires the Electron-based desktop VS Code.";
 
 export async function renderHtmlToPdf(
   html: string,
@@ -210,9 +213,7 @@ export async function renderHtmlToPdf(
     await loaded;
     const print = panel.webview.printToPDF;
     if (typeof print !== "function") {
-      throw new Error(
-        "[PDF-PRINT] webview.printToPDF is not available in this VS Code runtime. Requires the Electron-based desktop VS Code."
-      );
+      throw new Error(UNSUPPORTED_RUNTIME_MESSAGE);
     }
     return await print({
       marginsType: 0,
@@ -281,6 +282,9 @@ async function runExport(
 ): Promise<void> {
   log.info("export-pdf invoked", { uri: sourceUri.toString() });
   try {
+    if (deps.isPrintToPdfRuntimeSupported?.() === false) {
+      throw new Error(UNSUPPORTED_RUNTIME_MESSAGE);
+    }
     const t0 = Date.now();
     const src = await readMarkdown(sourceUri, deps);
     const title = titleFromPath(sourceUri.path);

--- a/packages/vscode/src/extension.ts
+++ b/packages/vscode/src/extension.ts
@@ -153,6 +153,7 @@ export const activate = (context: vscode.ExtensionContext) => {
       readFile: (u) => Promise.resolve(vscode.workspace.fs.readFile(u)).then((b) => new Uint8Array(b)),
       writeFile: (u, data) => Promise.resolve(vscode.workspace.fs.writeFile(u, data)),
       uriWithPath: (base, newPath) => base.with({ path: newPath }),
+      isPrintToPdfRuntimeSupported: () => vscode.env.uiKind === vscode.UIKind.Desktop,
       createWebviewPanel: (viewType, title, showOptions, opts) =>
         vscode.window.createWebviewPanel(
           viewType,

--- a/packages/vscode/test/export-pdf.test.ts
+++ b/packages/vscode/test/export-pdf.test.ts
@@ -456,6 +456,22 @@ describe("exportPdf composer", () => {
     expect(lines.some((l) => l.includes("export-pdf failed"))).toBe(true);
   });
 
+  it("preflights unsupported runtimes before reading or composing the markdown", async () => {
+    const { deps, spies } = makeDeps({ noPrintToPdf: true });
+    const unsupportedDeps: ExportPdfDeps = {
+      ...deps,
+      isPrintToPdfRuntimeSupported: () => false,
+    };
+
+    await exportPdf(SAMPLE_URI, { theme: "light" }, unsupportedDeps);
+
+    expect(spies.readFile).not.toHaveBeenCalled();
+    expect(spies.createWebviewPanel).not.toHaveBeenCalled();
+    expect(spies.writeFile).not.toHaveBeenCalled();
+    expect(spies.showErrorMessage).toHaveBeenCalledTimes(1);
+    expect((spies.showErrorMessage.mock.calls[0]?.[0] as string).toLowerCase()).toContain("desktop vs code");
+  });
+
   it("serialises concurrent invocations on the same URI (per-uri lock)", async () => {
     const { deps, spies } = makeDeps();
     const a = exportPdf(SAMPLE_URI, { theme: "light" }, deps);

--- a/packages/vscode/test/vscode-mock.ts
+++ b/packages/vscode/test/vscode-mock.ts
@@ -2,6 +2,7 @@
 import { vi } from "vitest";
 
 export const ViewColumn = { Beside: 2, Active: -1 } as const;
+export const UIKind = { Desktop: 1, Web: 2 } as const;
 
 const FAKE_PDF = new TextEncoder().encode("%PDF-1.7\n" + "x".repeat(2048) + "\n%%EOF\n");
 
@@ -124,5 +125,6 @@ export const Uri = {
 };
 
 export const env = {
+  uiKind: UIKind.Desktop,
   openExternal: vi.fn((_u: unknown) => Promise.resolve(true)),
 };

